consensus/ethash/progpow.go

```diff
@@ -0,0 +1,410 @@
+package ethash
+
+import (
+	"encoding/binary"
+	"github.com/ethereum/go-ethereum/crypto/sha3"
+)
+
+const (
+	progpowCacheWords = 4 * 1024 // Total size 16*1024 bytes
+	progpowLanes      = 32
+	progpowRegs       = 16
+	progpowCntCache   = 8
+	progpowCntMath    = 8
+	progpowCntMem     = loopAccesses
+	progpowMixBytes   = 2 * mixBytes
+)
+
+func progpowLight(size uint64, cache []uint32, hash []byte, nonce uint64,
+	blockNumber uint64) ([]byte, []byte) {
+	keccak512 := makeHasher(sha3.NewKeccak512())
+
+	lookup := func(index uint32) uint32 {
+		rawData := generateDatasetItem(cache, index/16, keccak512)
+		data := binary.LittleEndian.Uint32(rawData[(index%16)*4:])
+		return data
+	}
+	return progpow(hash, nonce, size, blockNumber, lookup)
+}
+
+func progpowFull(dataset []uint32, hash []byte, nonce uint64,
+	blockNumber uint64) ([]byte, []byte) {
+	lookup := func(index uint32) uint32 {
+		return dataset[index]
+	}
+	return progpow(hash, nonce, uint64(len(dataset))*4, blockNumber, lookup)
+}
+
+func rotl32(x uint32, n uint32) uint32 {
+	return (((x) << (n % 32)) | ((x) >> (32 - (n % 32))))
+}
+
+func rotr32(x uint32, n uint32) uint32 {
+	return (((x) >> (n % 32)) | ((x) << (32 - (n % 32))))
+}
+
+func lower32(in uint64) uint32 {
+	return uint32(in & uint64(0x00000000FFFFFFFF))
+}
+
+func higher32(in uint64) uint32 {
+	return uint32((in >> 32) & uint64(0x00000000FFFFFFFF))
+}
+
+var keccakfRNDC = [24]uint32{
+	0x00000001, 0x00008082, 0x0000808a, 0x80008000, 0x0000808b, 0x80000001,
+	0x80008081, 0x00008009, 0x0000008a, 0x00000088, 0x80008009, 0x8000000a,
+	0x8000808b, 0x0000008b, 0x00008089, 0x00008003, 0x00008002, 0x00000080,
+	0x0000800a, 0x8000000a, 0x80008081, 0x00008080, 0x80000001, 0x80008008}
+
+func keccakF800Round(st [25]uint32, r int) [25]uint32 {
+	var keccakfROTC = [24]uint32{1, 3, 6, 10, 15, 21, 28, 36, 45, 55, 2,
```

Comment:
These values are incorrect, probably copied from Keccak-f[1600]. You cannot rol 32-bit integer by 36 bits.

Reply:
I noticed those two. But it never occurred to me that the were copied from keccak1600. In the implementation (and in Go library math/bits), rolling by 36 bits means rolling by 4 bits :)

Reply:
In C this is undefined behavior, but somehow it works.